Library/Homebrew/cli/named_args.rb

```diff
@@ -29,13 +42,212 @@ def initialize(*args, parent: Args.new, override_spec: nil, force_bottle: false,
+        type = @type if only.blank?
+        type ||= case @type
+        when :formula_or_cask
+          only
+        when :installed_formula_or_cask
+          :"installed_#{only}"
+        else
+          @type
+        end
+
+        case type
+        when :formula
+          formulae += formulae ignore_unavailable: ignore_unavailable
+        when :installed_formula
+          formulae += formulae installed: true, ignore_unavailable: ignore_unavailable
+        when :cask
+          casks += casks ignore_unavailable: ignore_unavailable
+        when :installed_cask
+          casks += casks installed: true, ignore_unavailable: ignore_unavailable
+        when :formula_or_cask
+          formulae += formulae ignore_unavailable: true
+          casks += casks ignore_unavailable: true
+        when :installed_formula_or_cask
+          formulae += formulae installed: true, ignore_unavailable: true
+          casks += casks installed: true, ignore_unavailable: true
+        when :tap
+          taps += taps ignore_unavailable: ignore_unavailable
+        when :installed_tap
+          taps += taps installed: true, ignore_unavailable: ignore_unavailable
+        else
+          others += @args
+        end
+
+        # Display warning if name is both a formula and a cask
+        if !allow_formula_cask_conflicts && [:formula_or_cask, :installed_formula_or_cask].include?(type)
+          formula_names = formulae.map(&:name)
+          cask_tokens = casks.map(&:token)
+
+          (formula_names & cask_tokens).each do |conflict|
+            cask = casks.find { |c| c.token == conflict }
+            message = "Treating #{conflict} as a formula."
+            message += " For the cask, use #{cask.tap.name}/#{cask.token}" if cask.tap.present?
+            opoo message.freeze
+
+            casks.reject! { |c| c.token == conflict }
+          end
+        end
+
+        # Fail if there are extras and `ignore_unavailable` is `false`. Otherwise add to `others`.
+        installed_types = [:installed_formula, :installed_cask, :installed_formula_or_cask]
+        @args_to_convert.each do |name|
+          unless ignore_unavailable
+            raise FormulaOrCaskNotInstalledError, name if installed_types.include? type
+
+            raise FormulaOrCaskUnavailableError, name
+          end
+
+          others << name
+        end
+
+        if separate_types
+          { formula: formulae, cask: casks, keg: kegs, tap: taps, other: others }
+        else
+          formulae + casks + kegs + taps + others
+        end
+      end
+
+      # @api public
+      sig do
+        params(only: T.nilable(Symbol), ignore_unavailable: T::Boolean, allow_formula_cask_conflicts: T::Boolean)
+          .returns(T.any(Array, Hash))
+      end
+      def convert_to_paths(only: nil, ignore_unavailable: false, allow_formula_cask_conflicts: true)
+        paths = []
+        backup_args = @args
+
+        @args.select! do |arg|
```

Comment:
Don't use destructive methods. `backup_args` will actually be a just a reference and be changed as well.